net/dns/resolver/tsdns.go

```diff
@@ -708,6 +702,57 @@ func (r *Resolver) resolveLocal(domain dnsname.FQDN, typ dns.Type) (netaddr.IP,
 	}
 }
 
+// resolveMagic provides DNS resolution for a number of specialcase domains.
+//
+// The resolved IP and true is returned if resolveMagic has an authorative
+// answer for the query, or netaddr.IP{} + false otherwise.
+func (r *Resolver) resolveMagic(domain dnsname.FQDN, typ dns.Type) (netaddr.IP, bool) {
+	// We return a symbolic domain if someone does a reverse lookup on the
+	// DNS endpoint. To round out this special case, we also do the inverse
+	// (returning the endpoint IP if someone looks up the symbolic domain).
+	if domain == dnsSymbolicFQDN {
+		switch typ {
+		case dns.TypeA:
+			return tsaddr.TailscaleServiceIP(), true
+		case dns.TypeAAAA:
+			return tsaddr.TailscaleServiceIPv6(), true
+		}
+	}
+
+	// As a convenient mapping into Tailscales 'Via Range', we support
+	// resolving any queries of the form 'via-<X>.<IPv4-address>' into
+	// their corresponding IPv6 address, where X is a base 10 or base 16
+	// siteID.
+	fqdn := string(domain)
+	if firstDot := strings.Index(fqdn, "."); len(fqdn) > 10 && firstDot > 0 && strings.HasPrefix(fqdn, "via-") {
```

Comment:
that 10 is mysterious.  if the goal is to do something cheap early, it's kinda too late as you've already called strings.Index.

I'd pull this into a func where you can return early a bunch instead of making this a bunch of nested ifs indenting far.

```go 
...
      if typ == dns.TypeAAAA {
          if ip, ok := parseViaName(domain); ok {
                 return ip, true
          }
      }
      // TODO: other magic in the future?
      return  netaddr.IP{}, false
}
```

Then like:

```go
func parseViaName(domain dnsname.FQDN) (ip netaddr.IP, ok bool) {
     s := string(domain)
     if len(s) < len("via-X.0.0.0.0") {
            return
      }
      if !strings.HasPrefix(s, "via-") {
            return
      }
      s = s[len("via-"):]
      ....
```

(Note that `len("via-")`compiles to 4)
```

Reply:
Ahhhh thats so much nicer!